bundles/org.openhab.binding.melcloud/src/main/java/org/openhab/binding/melcloud/internal/discovery/MelCloudDiscoveryService.java

```diff
@@ -0,0 +1,178 @@
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.config.discovery.DiscoveryService;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.core.thing.binding.ThingHandler;
+import org.eclipse.smarthome.core.thing.binding.ThingHandlerService;
+import org.openhab.binding.melcloud.internal.MelCloudBindingConstants;
+import org.openhab.binding.melcloud.internal.api.json.Device;
+import org.openhab.binding.melcloud.internal.exceptions.MelCloudCommException;
+import org.openhab.binding.melcloud.internal.exceptions.MelCloudLoginException;
+import org.openhab.binding.melcloud.internal.handler.MelCloudAccountHandler;
+import org.osgi.service.component.annotations.Modified;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link MelCloudDiscoveryService} creates things based on the configured location.
+ *
+ * @author Luca Calcaterra - Initial Contribution
+ * @author Pauli Anttila - Refactoring
+ * @author Wietse van Buitenen - Check device type, added heatpump device
+ */
+public class MelCloudDiscoveryService extends AbstractDiscoveryService
+        implements DiscoveryService, ThingHandlerService {
+    private final Logger logger = LoggerFactory.getLogger(MelCloudDiscoveryService.class);
+
+    private static final int DISCOVER_TIMEOUT_SECONDS = 10;
+
+    private @NonNullByDefault({}) MelCloudAccountHandler melCloudHandler;
+    private ScheduledFuture<?> scanTask;
+
+    /**
+     * Creates a MelCloudDiscoveryService with enabled autostart.
+     */
+    public MelCloudDiscoveryService() {
+        super(MelCloudBindingConstants.DISCOVERABLE_THING_TYPE_UIDS, DISCOVER_TIMEOUT_SECONDS, true);
+    }
+
+    @Override
+    protected void activate(Map<String, @Nullable Object> configProperties) {
+        super.activate(configProperties);
+    }
+
+    @Override
+    public void deactivate() {
+        super.deactivate();
+    }
+
+    @Override
+    @Modified
+    protected void modified(Map<String, @Nullable Object> configProperties) {
+        super.modified(configProperties);
+    }
+
+    @Override
+    protected void startBackgroundDiscovery() {
+        discoverDevices();
+    }
+
+    @Override
+    protected void startScan() {
+        if (this.scanTask != null) {
+            scanTask.cancel(true);
+        }
+        this.scanTask = scheduler.schedule(() -> discoverDevices(), 0, TimeUnit.SECONDS);
+    }
+
+    @Override
+    protected void stopScan() {
+        super.stopScan();
+
+        if (this.scanTask != null) {
+            this.scanTask.cancel(true);
+            this.scanTask = null;
+        }
+    }
+
+    private void discoverDevices() {
+        logger.debug("Discover devices");
+
+        if (melCloudHandler != null) {
```

Comment:
By using the annotation which you used on line 54 we normally indicate that the framework guarantees the field to be set when needed. It does not match with the null-check here. Does the framework guarantee it to be set? If not please remove the annotation on line 54. If yes, remove his `null `check.

Reply:
This comment still applies, it would be nice to have resolved it will now form a bad example for the next OH developer.